### ISSUE TITLE
Allow Offset to be Changed with Time Grouping Change

### ISSUE THREAD

**Issue by ixipixi** [metabase/metabase#51167]
**Is your feature request related to a problem? Please describe.**
In my dashboard I would like to use it with a simple trend chart that shows the total orders in our shop. It should compare the current (daily or weekly) number to the previous period (previous day or week) and to the previous year (same day or week last year). The comparison to last period is no problem as this changes according to the time grouping filter. But for the YoY comparison I used the offset function -> for daily base I used -366 days. 

If we use the time grouping parameter to change the granularity the offset no longer make sense.

**Describe the solution you'd like**
Some way to convert the offset to the appropriate value or set a different offset for different time granularities.

**How important is this feature to you?**
Important - there isn't a way to do use these in conjunction in a way that makes sense.

**Comment by thebiglabasky** [comment]
Is it fair to say that the Window function is the technical way to do the query yet not the thing people actually care about: they want to compare something with a given equivalent in the past and that's it right?

**Comment by ixipixi** [comment]
@thebiglabasky yes. They want the trend to always compare to an equivalent period.

**Comment by mngr** [comment]
Yes, we don't have a good solution for this atm, but it's in my backlog

**Comment by cedricoeldorf** [comment]
It would help if you could expose the current time grouping (e.g. Hour, Day, Week, Month, etc.) as a variable that we can reference in custom expressions. That way, we could pass the user’s selected granularity directly into the offset() function with a case() statement.

`offset(SUM([Orders]), case([Time Grouping] = "Week", -52, [Time Grouping] = "Month", -12, ...))`

**Comment by cedricoeldorf** [comment]
Additionally, it’s clear that Offset is often used to calculate timestamp-based growth or change (e.g. YoY comparisons). 

But Metabase currently doesn’t support filling missing values in time series. So if, say, June 2023 has no data, the offset() ends up comparing June 2024 to April 2024, which breaks the entire logic.

These two issues are tightly linked and really need to be solved together:
1) Accurate time-aware offset based on the selected grain
2) Optional fill of missing date periods for time-based consistency

Without both, time-based growth charts in Metabase remain fragile and misleading.

**Comment by thebiglabasky** [comment]
MetricFlow have a concept of time spine to solve for the missing date periods problem

**Comment by shivampaw** [comment]
+1 to this.

I've wrote a Calendar DIM which has grains for "Series (Current/Previous), "Day", "Week", "Month", & "Quarter" and join my data onto it…twice…by the fact_dt and display_dt and oh my lord it's becoming impossible to work with…

Being able to set a different offset based on time grouping would go a long way to solving this…but it also needs to take into account leap years / 53 weeks in a year possibility...ideally

**Comment by nasu-ndg** [comment]
+1
